044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,154 @@
+for a Kafka cluster that is using KRaft rather than ZooKeeper. This includes a KRaft 
+cluster with combined nodes and one that has separate controller and broker nodes.
+
+## Current situation
+
+These are the current probes used in Strimzi when ZooKeeper is present:
+
+|Component| Liveness | Readiness |
+|---|---|---|
+| ZooKeeper | "imok" from 127.0.0.1:12181 | "imok" from 127.0.0.1:12181 |
+| ZK-based broker | if (ready) { listening on 9091 } else { have ZK session } | if (ready) { return true } |
+
+The "ready" check in both cases is based on the existence of the file `../kafka-ready` which is 
+created by the KafkaAgent when the broker state metric has value 3 (RUNNING).
+
+## Background
+
+This section contains some background information about the ports and metrics we could use for checks.
+
+### Ports
+In KRaft mode there are two different ports that the node could be listening on.
+If it has a controller process it should be listening on the port defined as the `controller.listener.names` which in Strimzi is hard-coded to 9090.
+If it has a broker process it should be listening on the replication port specified as the `inter.broker.listener.name` which is hard-coded to 9091.
+If the node is running as a combined mode, it should be listening on both ports
+
+### BrokerState
+The expected [BrokerState](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/BrokerState.java) transitions in the "happy path" during a normal broker lifecycle are:
+
+NOT_RUNNING (0) -> STARTING (1) -> RECOVERY (2) -> RUNNING (3) -> PENDING_CONTROLLED_SHUTDOWN (6) -> SHUTTING_DOWN (7)
+
+The other possible value is 127 which represents UNKNOWN.
+
+### Current state metric and quorum
+
+Some useful things to be aware of with the current-state metric and how the controller quorum affects the broker nodes:
+
+* The [current-state](https://kafka.apache.org/documentation/#kraft_quorum_monitoring) metric indicates whether a controller is currently a leader or follower.
+If this metric shows one of these two states, this means the leader election has happened and a controller quorum has been successfully formed.
+* The current-state metric is not accessible as a Yammer metric, so cannot be accessed by the KafkaAgent in the same way it currently reads the BrokerState metric.
+* The current-state metric could perhaps be read by the KafkaAgent if it called the MBean server within the JVM directly or read it from `$LOG_DIR/__cluster_metadata-0/quorum-state`.
+* The broker nodes will not move out of the STARTING state until the quorum leader election has happened.
+
+## Proposal
+
+There are three different types of liveness/readiness probe that we need to consider, first for a controller only, then 
+for a broker only, and finally for a combined node that has both roles.
+
+The following statements describe the intent of the proposed probes:
+
+* A controller is:
+    * "alive" if it has a process running.
```

Comment:
If this is really the check we want why not use `ps`? Or an HTTP server running in the server for which the `bind`/`listen`/`accept` doesn't depend deeply on anything about the Kafka code?

Reply:
I think we should mention this is the "minimal exception", not exactly what we will do. Otherwise, it's a little confusing when seeing "we just want to check if it has a process running" here, and then seeing "we check controller is listening on the port" below.

Reply:
I'm not sure what you mean by "minimal exception". I put above that these statements "describe the intent of the proposed probes". So the intent is that the controller should be marked alive if there is some sort of process running, then lower down we describe how we actually work out that a process is running. So that could be by seeing that it is listening on a port, or as Tom says by a basic `ps` check or running some other process in the container.